src/cluster_migrateslots.c

```diff
@@ -0,0 +1,1954 @@
+            setSlotImportingStateInAllDbs(job->slot_ranges, 0);
+            updateSlotMigrationJobState(job, job->post_cleanup_state);
+            return;
+
+        /* Exporting states */
+        case SLOT_EXPORT_CONNECTING:
+            if (!job->conn) {
+                connectSlotExportJob(job);
+            }
+            if (proceedWithSlotExportJobConnecting(job)) {
+                continue;
+            }
+            return;
+        case SLOT_EXPORT_AUTHENTICATING:
+            performSlotExportJobAuthentication(job);
+            continue;
+        case SLOT_EXPORT_SEND_ESTABLISH:
+            initSlotExportJobClient(job);
+            addReplySds(job->client, generateSyncSlotsEstablishCommand(job));
+            connSetReadHandler(job->client->conn, slotMigrationJobReadEstablishResponse);
+            updateSlotMigrationJobState(job, SLOT_EXPORT_READ_ESTABLISH_RESPONSE);
+            return;
+        case SLOT_EXPORT_READ_ESTABLISH_RESPONSE:
+            /* We are still reading back the response, nothing to do in cron */
+            return;
+        case SLOT_EXPORT_WAITING_TO_SNAPSHOT:
+            /* Perform the snapshot whenever there is no child process.
+             *
+             * We also check that there are no pending writes here. Since we are
+             * sending ACKs at this time, we could have unfortunate timing where
+             * an ACK is added to the output buffer just before snapshotting,
+             * adding the client to the pending write queue. This write might be
+             * flushed after some other command is enqueued during the snapshot,
+             * resulting in premature flush of the output buffer and data
+             * consistency issues. To prevent this, we defer snapshot until
+             * there are no pending writes. */
+            if (hasActiveChildProcess() || job->client->flag.pending_write) {
+                return;
+            }
+            slotExportJobBeginSnapshot(job);
+            return;
+        case SLOT_EXPORT_SNAPSHOTTING:
+            /* Waiting for child process to finish */
+            return;
+        case SLOT_EXPORT_STREAMING:
+            /* Waiting for PAUSE command to come in */
+            return;
+        case SLOT_EXPORT_WAITING_TO_PAUSE:
+            slotExportTryDoPause(job);
+            return;
+        case SLOT_EXPORT_FAILOVER_PAUSED:
+            if (isSlotExportPauseTimedOut(job)) {
+                serverLog(LL_WARNING, "Slot migration %s timed out during slot failover.", job->description);
+                job->mf_end = 0;
+                updatePausedActions();
+                finishSlotMigrationJob(job, SLOT_MIGRATION_JOB_FAILED, "Timed out before streaming completed");
+            }
+            return;
+        case SLOT_EXPORT_FAILOVER_GRANTED:
+            if (isSlotExportPauseTimedOut(job)) {
+                /* Note that we think this won't happen very commonly. The main source of
```

Comment:
"not very commonly" is not a strong guarantee nor very informative. this error will happen in the real world so I think it would be really helpful to give a few examples, such as transient network issues/partitioning.